### PR TITLE
2D Field View Wrappers

### PR DESCRIPTION
**Pull request by sslattery**
Adding field view wrappers for 2D grids.

### CODE REVIEW

File: src/Picasso_FieldTypes.hpp [hunk 1]
```diff
@@ -287,19 +289,34 @@ struct ScalarViewWrapper
...
+    {
+        return _v( i0, i1, 0 );
+    }
+
     // Access the view data through full index arguments.
-    KOKKOS_FORCEINLINE_FUNCTION
-    value_type& operator()( const int i0, const int i1, const int i2,
-                            const int ) const
+    template <int VR = view_rank>
+    KOKKOS_FORCEINLINE_FUNCTION std::enable_if_t<4 == VR, value_type&>
```
Comment (kwitaechong) [inline src/Picasso_FieldTypes.hpp:308]: any difference from the `enable_if_t<4 == VR, linear_algebra_type&>`?

File: src/Picasso_FieldTypes.hpp [hunk 2]
```diff
@@ -389,21 +429,36 @@ struct TensorViewWrapper
...
     // layout-left where the dim0 and dim1 dimensions are switched.
-    KOKKOS_FORCEINLINE_FUNCTION
-    LinearAlgebra::MatrixView<value_type, dim0, dim1>
+    template <int VR = view_rank>
+    KOKKOS_FORCEINLINE_FUNCTION std::enable_if_t<4 == VR, linear_algebra_type>
     operator()( const int i0, const int i1, const int i2 ) const
     {
-        return LinearAlgebra::MatrixView<value_type, dim0, dim1>(
-            &_v( i0, i1, i2, 0 ), dim1 * _v.stride( 3 ), _v.stride( 3 ) );
+        return linear_algebra_type( &_v( i0, i1, i2, 0 ), dim1 * _v.stride( 3 ),
```
Comment (kwitaechong) [inline src/Picasso_FieldTypes.hpp:436]: if tensor components are unrolled in the last dimension, `_v.stride(3) = dim0*dim1`? Then is correct way to access?

File: src/Picasso_FieldTypes.hpp [hunk 3]
```diff
@@ -333,21 +355,34 @@ struct VectorViewWrapper
...
     // Access the view data as a vector through point-wise index arguments.
-    KOKKOS_FORCEINLINE_FUNCTION
-    LinearAlgebra::VectorView<value_type, dim0>
+    template <int VR = view_rank>
+    KOKKOS_FORCEINLINE_FUNCTION std::enable_if_t<4 == VR, linear_algebra_type>
     operator()( const int i0, const int i1, const int i2 ) const
     {
-        return LinearAlgebra::VectorView<value_type, dim0>(
-            &_v( i0, i1, i2, 0 ), _v.stride( 3 ) );
+        return linear_algebra_type( &_v( i0, i1, i2, 0 ), _v.stride( 3 ) );
```
Comment (kwitaechong) [inline src/Picasso_FieldTypes.hpp:362]: would this access to vector size of dim0 at (i0,i1,i2)?

File: src/Picasso_FieldTypes.hpp (hunk 2 above)
Comment (sslattery) [reply src/Picasso_FieldTypes.hpp:436]: The stride refers to the stride between field components in the Kokkos view. So here `_v.stride(3)` is either `1` if `Kokkos::LayoutLeft` or `NumI * NumJ * NumK` if `Kokkos::LayoutRight`

Comment (sslattery) [reply src/Picasso_FieldTypes.hpp:436]: The stride refers to the stride between the first component of each tensor in the Kokkos view. So here `_v.stride(3)` is either `1` if `Kokkos::LayoutLeft` or `NumI * NumJ * NumK` if `Kokkos::LayoutRight`

File: src/Picasso_FieldTypes.hpp (hunk 3 above)
Comment (sslattery) [reply src/Picasso_FieldTypes.hpp:362]: See my response above - this is the stride between the first component of each vector in the Kokkos view.

File: src/Picasso_FieldTypes.hpp (hunk 2 above)
Comment (kwitaechong) [reply src/Picasso_FieldTypes.hpp:436]: So for 2D, (i=0,...,NumI), (j=0,...,NumJ) and tensor (M,N),  then `_v` is stored as first tensor component is stored at the grid NumI*NumJ and the second component of tensor is stored at the grid NumI*NumJ?

**Review by kwitaechong — APPROVED** [review]

**Pull request merged**